src/lib.rs

```diff
@@ -1,41 +1,600 @@
+        let new_id = self.lend_data.new_deposit();
+        let ctokens_amount = self.count_ctokens(amount);
+
+        let new_deposit = Payment {
+            amount,
+            ctokens_amount,
+            interest_rate: self.lend_data.interest_rate,
+            payment_time: exec::block_timestamp() / 1000, // time in secs
+            id: new_id,
+        };
+
+        if balance(&self.token_address, &msg::source()).await < amount {
+            panic!("Not enough balance")
+        }
+
+        transfer_tokens(
+            &self.token_address,
+            &msg::source(),
+            &self.owner_address,
+            amount,
+        )
+        .await;
+        transfer_tokens(
+            &self.ctoken_address,
+            &self.owner_address,
+            &msg::source(),
+            ctokens_amount,
+        )
+        .await;
+
+        if Compound::insert_payment(msg::source(), &mut self.user_deposits, &new_deposit) {
+            self.lend_data.remove_repeated()
+        }
+
+        self.user_assets
+            .entry(msg::source())
+            .and_modify(|(lend, _)| *lend += new_deposit.amount)
+            .or_insert_with(|| (new_deposit.amount, 0));
+    }
+
+    pub async fn borrow_tokens(&mut self, amount: u128) {
+        asserts::greater_zero(amount, "Borrow token amount");
+
+        let (user_lend, user_borrow) = self
+            .user_assets
+            .get(&msg::source())
+            .expect("No user assets found");
+
+        if user_lend - user_borrow < amount {
+            panic!(
+                "Not possible to borrow {} tokens due to the collateral factor",
+                amount
+            );
+        }
+
+        transfer_tokens(
+            &self.token_address,
+            &self.owner_address,
+            &msg::source(),
+            amount,
+        )
+        .await;
+
+        let new_id = self.borrow_data.new_borrow();
+        let new_borrow = Payment {
+            amount,
+            ctokens_amount: 0,
+            interest_rate: self.borrow_data.borrow_rate,
+            payment_time: exec::block_timestamp() / 1000, // time in secs
+            id: new_id,
+        };
+
+        if Compound::insert_payment(msg::source(), &mut self.user_borrows, &new_borrow) {
+            self.borrow_data.remove_repeated()
+        }
+
+        self.user_assets
+            .entry(msg::source())
+            .and_modify(|(_, borrow)| *borrow += new_borrow.amount);
+    }
+
+    pub async fn redeem_tokens(
+        &mut self,
+        lend_to_redeem: BTreeSet<u128>,
+        amount_per_borrow: BTreeMap<u128, u128>,
+    ) {
+        if !self.redeem_posible(msg::source(), &lend_to_redeem, &amount_per_borrow) {
+            panic!("Redeem such amount of tokens is not possible due to collateral factor")
+        }
+
+        let mut borrow_iter = amount_per_borrow.iter();
+
+        let mut redeemed_deposits = BTreeSet::<Payment>::new();
+        let mut redeem_borrowed = BTreeMap::<u128, Payment>::new();
```

Comment:
No need to set types explicitly when they can be inferred automatically.
```suggestion
        let mut redeemed_deposits = BTreeSet::new();
        let mut redeem_borrowed = BTreeMap::new();
```